### PR TITLE
Add begining image function and demo and performance optimizations (10x improvements related to logging level skips)

### PR DESCRIPTION
**Pull request by ldemailly**
./grol examples/image.gr

Produces

Time elapsed:  0.6636569499969482  seconds
Saved image to grol.png

grol

### CODE REVIEW

File: extensions/images.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		Help:       "create a new image of the name and size",
+		ArgTypes:   []object.Type{object.STRING, object.INTEGER, object.INTEGER},
+		ClientData: cdata,
+		Callback: func(cdata any, _ string, args []object.Object) object.Object {
+			images := cdata.(ImageMap)
+			x := int(args[1].(object.Integer).Value)
+			y := int(args[2].(object.Integer).Value)
+			if x > MaxImageSize || y > MaxImageSize {
+				return object.Error{Value: "image size too large"}
+			}
```
Comment (ccoVeille) [inline extensions/images.go:73]: I had to find this to understand MaxImageSize was in pixel and not bytes

Maybe a name with pixel in it

MaxPixel or MaxImagePixel

File: extensions/images.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+			img := image.NewRGBA(image.Rect(0, 0, x, y))
+			transparent := color.RGBA{0, 0, 0, 0}
+			draw.Draw(img, img.Bounds(), &image.Uniform{transparent}, image.Point{}, draw.Src)
+			images[args[0]] = img
+			return args[0]
+		},
+	}
+	MustCreate(imgFn)
+	imgFn.Name = "image_set"
+	imgFn.Help = "set a pixel in the named image, color is an array of 3 or 4 elements 0-255"
```
Comment (ccoVeille) [inline extensions/images.go:86]: You should mention RGB and RGBA people might not get what are these 3-4 number

Also I think most people are used to define color in hexa, not in decimal

File: extensions/images.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	imgFn.MinArgs = 4
+	imgFn.MaxArgs = 4
+	imgFn.ArgTypes = []object.Type{object.STRING, object.INTEGER, object.INTEGER, object.ARRAY}
+	imgFn.Callback = func(cdata any, _ string, args []object.Object) object.Object {
+		images := cdata.(ImageMap)
+		x := int(args[1].(object.Integer).Value)
+		y := int(args[2].(object.Integer).Value)
+		img, ok := images[args[0]]
```
Comment (ccoVeille) [inline extensions/images.go:94]: Please define a variable for args[0] and args[3]

It's unclear what they are

File: extensions/images.go (hunk 2 above)
Comment (ldemailly) [reply extensions/images.go:86]: so 0x00-0xFF instead? not sure that's clearer, I was thinking of passing as a single int but I think it's cleaner as triplet - will mention alpha channel

File: extensions/images.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			transparent := color.RGBA{0, 0, 0, 0}
+			draw.Draw(img, img.Bounds(), &image.Uniform{transparent}, image.Point{}, draw.Src)
+			images[args[0]] = img
+			return args[0]
+		},
+	}
+	MustCreate(imgFn)
+	imgFn.Name = "image_set"
+	imgFn.Help = "set a pixel in the named image, color is an array of 3 or 4 elements 0-255"
+	imgFn.MinArgs = 4
```
[the remark this answers is not in the document]
Comment (ldemailly) [reply extensions/images.go:87]: color isn't optional, it's in the array of color that it can be 3 or 4

File: extensions/images.go (hunk 2 above)
Comment (ccoVeille) [reply extensions/images.go:86]: you might be right rgba seems to commonly be used with https://www.w3schools.com/css/css3_colors.asp

https://rgbacolorpicker.com/rgba-to-hex

maybe you could keep the current format and accept one param when it starts with #

File: extensions/images.go (hunk 1 above)
Comment (ldemailly) [reply extensions/images.go]: changed

File: extensions/images.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		}
+		outputFile, err := os.Create("grol.png")
+		if err != nil {
+			return object.Error{Value: "error opening image file: " + err.Error()}
+		}
+		err = png.Encode(outputFile, img)
+		if err != nil {
+			return object.Error{Value: "error encoding image: " + err.Error()}
+		}
+		outputFile.Close()
```
[the remark this answers is not in the document]
Comment (ldemailly) [reply extensions/images.go]: done ty (even though I don't like defer it's appropriate here (even though I doubt png.Encode ever errors out or if it does that closing the file does anything but should avoid leaks

File: extensions/images.go (hunk 3 above)
Comment (ldemailly) [reply extensions/images.go:95]: added comment about args[0] being always the image ref/name

args[3] is use directly in a line that has "color" and "array" :)

File: extensions/images.go (hunk 2 above)
Comment (ldemailly) [reply extensions/images.go]: done (adding the other color modes)

**Review by ccoVeille — APPROVED** [review]

**Pull request merged**